### ISSUE TITLE
How to access the index of element in RuleForEach WithMessage when using SetValidator?

### ISSUE THREAD

**Issue by rmlira** [FluentValidation/FluentValidation#1293]
### System Details

- FluentValidation version: 8.6.1

### Issue Description

How to access the index of element in RuleForEach WithMessage when using SetValidator?
As an example:

```
RuleFor(x => x.Orders)
 .NotEmpty()
 .WithMessage("Orders list must not be empty")
 .ForEach(order => {
   order.SetValidator(new OrderValidator());
 });

...

public class OrderValidator: AbstractValidator<Order>
{
  public OrderValidator()
  {
     RuleFor(order => order.Number)
       .NotEmpty()
       .WithMessage("{PropertyName} at {CollectionIndex} must not be empty");
  }
}
```

Is there a way to access parent context collection index?

Related to https://github.com/JeremySkinner/FluentValidation/issues/1166

Thanks!

**Comment by rmlira** [comment]
#1166

**Comment by rmlira** [comment]
#1166

**Comment by JeremySkinner** [comment]
The CollectionIndex placeholder isn't available in child validators like this as child validators are executed in isolation from the parent - they have no knowledge of where the object being validated came from or whether it's part of a collection or not (this is by design). 

CollectionIndex is only available when using `RuleForEach` directly against a collection element (not against the element's properties unfortunately - it's too far down the chain at this point), so you'd have to do it like this:

```csharp
RuleForEach(x => x.Orders)
   .Must(order => order.Number != null)
   .WithMessage("{PropertyName} at {CollectionIndex}");
```

But even that won't be exactly what you're after as the `{PropertyName}` placeholder will be filled in as "Orders" rather than "Orders[0].ProductName". 

Basically {CollectionIndex} is only useful when working with simple collections (eg of strings), but isn't particularly useful when trying to access individual properties of collection elements. It's something that we might be able to enhance in the future, but that's probably quite a way off I'm afraid.

**Comment by JeremySkinner** [comment]
Added in 9.0-preview3.
